tests/test_force_merge_service.py

```diff
@@ -0,0 +1,39 @@
+import sys
+
+import pytest
+from elasticsearch.exceptions import ConnectionTimeout
+
+from gdcmodels.esutils import force_merge_elasticsearch_indices
+
+if sys.version_info > (3, 2):
+    from unittest.mock import Mock
+
+
+@pytest.mark.parametrize("index", ["", None, [], 1])
+def test_force_merge__index_has_to_be_list_and_not_empty(es, index):
+    with pytest.raises(ValueError):
+        force_merge_elasticsearch_indices(es, index)
+
+
+@pytest.mark.skipif(sys.version_info < (3, 5), reason="magicmock require version 3.3+")
+def test_force_merge__finished_in_time(es):
+    index = ["test_index"]
+    es.indices.forcemerge = Mock(return_vallue={})
+    force_merge_elasticsearch_indices(es, index)
+    es.indices.forcemerge.asssert_called_once_with(index, 1)
+
+
+@pytest.mark.skipif(sys.version_info < (3, 5), reason="magicmock require version 3.3+")
+def test_force_merge__timeout(es):
+    index = ["test_index"]
+    es.indices.forcemerge = Mock(side_effect=ConnectionTimeout())
+    counts = [1, 0]
+    value = [
+        {"nodes": {"test_node": {"thread_pool": {"force_merge": {"active": count}}}}}
+        for count in counts
+    ]
+    es.nodes.stats = Mock(side_effect=value)
+    force_merge_elasticsearch_indices(es, index)
+    es.indices.forcemerge.asssert_called_once_with(index, 1)
+    es.nodes.stats.assert_called_with(metric="thread_pool")
+    assert es.nodes.stats.call_count == 2
```

Comment:
I really discourage this style of mocking in unit tests.

The implementation of the test is tightly coupled to the implementation of the function.   This means that we could change the implementation of the function while maintaining the same behavior, but we'd still have to update the unit test. This means that the unit test is not testing behavior.  Instead, it's testing how the function is implemented.

Also, this test is going to incur the sleep time, which will make the tests slow.  We should avoid this in unit tests.

I recommend omitting these two tests altogether since neither of them actually verify that a force merge happened.  Instead, test the script by hand to verify that this function works as intended.

Reply:
I was going to say that we could set up integration tests to verify this, and try to adjust the ES settings to encourage a proliferation of segments, but I'm not sure that'd be worth the effort, and we'll need to hand-test the script to verify it anyway.

Reply:
this is something I suggested. are you saying that unit tests supposed to verify that the actual force merge happened? I don't think that's the purpose of the unit tests? Making sure that the code behaves as expected based on a different behavior of the external service is what we're testing here. If (for example) postgres changed its query implementation and we had it mocked with a different behavior, we'd have to update our tests in that case too, so I don't see a problem with this.